Falcon: register clock for SPI

--- a/arch/mips/lantiq/falcon/sysctrl.c
+++ b/arch/mips/lantiq/falcon/sysctrl.c
@@ -62,6 +62,7 @@
 #define ACTS_PADCTRL2	0x00200000
 #define ACTS_PADCTRL3	0x00200000
 #define ACTS_PADCTRL4	0x00400000
+#define ACTS_SSC0	0x00002000
 
 #define sysctl_w32(m,val,reg)	ltq_w32((val), sysctl_membase[m] + (reg))
 #define sysctl_r32(m,reg)	ltq_r32(sysctl_membase[m] + (reg))
@@ -247,6 +248,7 @@ void __init ltq_soc_init(void)
 	clkdev_add_sys("1e100c00.serial", SYSCTL_SYS1, ACTS_ASC0_ACT);
 	clkdev_add_sys("1e100d00.spi", SYSCTL_SYS1, ACTS_SSC0);
 	clkdev_add_sys("1e200000.i2c", SYSCTL_SYS1, ACTS_I2C_ACT);
+	clkdev_add_sys("1e100d00.spi", SYSCTL_SYS1, ACTS_SSC0);
 }
 
 
